### PR TITLE
net: set dhclient lease and pid files

### PR DESCRIPTION
**Pull request by holmanb**
```
net: set dhclient lease and pid files

This commit drops the sandboxing requirement by
invoking dhclient using pid and lease files that
apparmor already allows[1].

This is desirable as relocating the dhclient binary
has led to some bug fixes[2].

[1] dhclient's apparmor profile allow writing to:

/{,var/}run/dhclient*.pid lrw,
/{,var/}run/dhclient*.lease* lrw,

[2] fix hashes:

db86753f81af
919e22dc1d77
```

## Local Test Steps

<details>
<summary>I have been testing locally in LXD on jammy by forcing a call to dhclient during init-local (patch and logs in the dropdown).
</summary>
<br>

```
diff --git a/cloudinit/sources/DataSourceLXD.py b/cloudinit/sources/DataSourceLXD.py
index 34e4e00eb..a76c4cda0 100644
--- a/cloudinit/sources/DataSourceLXD.py
+++ b/cloudinit/sources/DataSourceLXD.py
@@ -24,6 +24,7 @@ from urllib3.connectionpool import HTTPConnectionPool
 
 from cloudinit import log as logging
 from cloudinit import sources, subp, util
+from cloudinit.net.dhcp import maybe_perform_dhcp_discovery
 
 LOG = logging.getLogger(__name__)
 
@@ -153,6 +154,7 @@ class DataSourceLXD(sources.DataSource):
         if not self._is_platform_viable():
             LOG.debug("Not an LXD datasource: No LXD socket found.")
             return False
+        maybe_perform_dhcp_discovery()
 
         self._crawled_metadata = util.log_time(
             logfunc=LOG.debug,
diff --git a/tests/unittests/sources/test_lxd.py b/tests/unittests/sources/test_lxd.py
index e60bb71fa..d85302c01 100644
--- a/tests/unittests/sources/test_lxd.py
+++ b/tests/unittests/sources/test_lxd.py
@@ -165,49 +165,49 @@ class TestDataSourceLXD:
     def test_subplatform(self, lxd_ds):
         assert "LXD socket API v. 1.0 (/dev/lxd/sock)" == lxd_ds.subplatform
 
-    def test__get_data(self, lxd_ds):
-        """get_data calls read_metadata, setting appropiate instance attrs."""
-        assert UNSET == lxd_ds._crawled_metadata
-        assert UNSET == lxd_ds._network_config
-        assert None is lxd_ds.userdata_raw
-        assert True is lxd_ds._get_data()
-        assert LXD_V1_METADATA == lxd_ds._crawled_metadata
-        # network-config is dumped from YAML
-        assert NETWORK_V1 == lxd_ds._network_config
-        # Any user-data and vendor-data are saved as raw
-        assert LXD_V1_METADATA["user-data"] == lxd_ds.userdata_raw
-        assert LXD_V1_METADATA["vendor-data"] == lxd_ds.vendordata_raw
-
-    def test_network_config_when_unset(self, lxd_ds):
-        """network_config is correctly computed when _network_config and
-        _crawled_metadata are unset.
-        """
-        assert UNSET == lxd_ds._crawled_metadata
-        assert UNSET == lxd_ds._network_config
-        assert None is lxd_ds.userdata_raw
-        # network-config is dumped from YAML
-        assert NETWORK_V1 == lxd_ds.network_config
-        assert LXD_V1_METADATA == lxd_ds._crawled_metadata
-
-    def test_network_config_crawled_metadata_no_network_config(
-        self, lxd_ds_no_network_config
-    ):
-        """network_config is correctly computed when _network_config is unset
-        and _crawled_metadata does not contain network_config.
-        """
-        lxd.generate_fallback_network_config = mock.Mock(
-            return_value=NETWORK_V1
-        )
-        assert UNSET == lxd_ds_no_network_config._crawled_metadata
-        assert UNSET == lxd_ds_no_network_config._network_config
-        assert None is lxd_ds_no_network_config.userdata_raw
-        # network-config is dumped from YAML
-        assert NETWORK_V1 == lxd_ds_no_network_config.network_config
-        assert (
-            LXD_V1_METADATA_NO_NETWORK_CONFIG
-            == lxd_ds_no_network_config._crawled_metadata
-        )
-        assert 1 == lxd.generate_fallback_network_config.call_count
+#    def test__get_data(self, lxd_ds):
+#        """get_data calls read_metadata, setting appropiate instance attrs."""
+#        assert UNSET == lxd_ds._crawled_metadata
+#        assert UNSET == lxd_ds._network_config
+#        assert None is lxd_ds.userdata_raw
+#        assert True is lxd_ds._get_data()
+#        assert LXD_V1_METADATA == lxd_ds._crawled_metadata
+#        # network-config is dumped from YAML
+#        assert NETWORK_V1 == lxd_ds._network_config
+#        # Any user-data and vendor-data are saved as raw
+#        assert LXD_V1_METADATA["user-data"] == lxd_ds.userdata_raw
+#        assert LXD_V1_METADATA["vendor-data"] == lxd_ds.vendordata_raw
+
+#    def test_network_config_when_unset(self, lxd_ds):
+#        """network_config is correctly computed when _network_config and
+#        _crawled_metadata are unset.
+#        """
+#        assert UNSET == lxd_ds._crawled_metadata
+#        assert UNSET == lxd_ds._network_config
+#        assert None is lxd_ds.userdata_raw
+#        # network-config is dumped from YAML
+#        assert NETWORK_V1 == lxd_ds.network_config
+#        assert LXD_V1_METADATA == lxd_ds._crawled_metadata
+
+#    def test_network_config_crawled_metadata_no_network_config(
+#        self, lxd_ds_no_network_config
+#    ):
+#        """network_config is correctly computed when _network_config is unset
+#        and _crawled_metadata does not contain network_config.
+#        """
+#        lxd.generate_fallback_network_config = mock.Mock(
+#            return_value=NETWORK_V1
+#        )
+#        assert UNSET == lxd_ds_no_network_config._crawled_metadata
+#        assert UNSET == lxd_ds_no_network_config._network_config
+#        assert None is lxd_ds_no_network_config.userdata_raw
+#        # network-config is dumped from YAML
+#        assert NETWORK_V1 == lxd_ds_no_network_config.network_config
+#        assert (
+#            LXD_V1_METADATA_NO_NETWORK_CONFIG
+#            == lxd_ds_no_network_config._crawled_metadata
+#        )
+#        assert 1 == lxd.generate_fallback_network_config.call_count
 
 
 class TestIsPlatformViable:
```

a log snippet of the results:
```
2022-09-02 02:16:12,927 - dhcp.py[DEBUG]: Performing a dhcp discovery on enp5s0
2022-09-02 02:16:12,927 - subp.py[DEBUG]: Running command ['ip', 'link', 'set', 'dev', 'enp5s0', 'up'] with allowed return codes [0] (shell=False, capture=True)
2022-09-02 02:16:12,935 - subp.py[DEBUG]: Running command ['/usr/sbin/dhclient', '-1', '-v', '-lf', '/run/dhclient.lease', '-pf', '/run/dhclient.pid', 'enp5s0', '-sf', '/bin/true'] with allowed return codes [0] (shell=False, capture=True)
2022-09-02 02:16:12,969 - util.py[DEBUG]: All files appeared after 0 seconds: ['/run/dhclient.pid', '/run/dhclient.lease']
2022-09-02 02:16:12,970 - util.py[DEBUG]: Reading from /run/dhclient.pid (quiet=False)
2022-09-02 02:16:12,970 - util.py[DEBUG]: Read 5 bytes from /run/dhclient.pid
2022-09-02 02:16:12,970 - util.py[DEBUG]: Reading from /proc/2743/stat (quiet=True)
2022-09-02 02:16:12,970 - util.py[DEBUG]: Read 312 bytes from /proc/2743/stat
2022-09-02 02:16:12,971 - dhcp.py[DEBUG]: killing dhclient with pid=2743
2022-09-02 02:16:12,973 - util.py[DEBUG]: Reading from /run/dhclient.lease (quiet=False)
2022-09-02 02:16:12,973 - util.py[DEBUG]: Read 1082 bytes from /run/dhclient.lease
```

</details>

## Cloud Test Steps

I grabbed the `cloud-init.log` files while manually testing on the following distros. The test was basically:
```
apt install ./cloud-init_22.3-21-g8c86e1f3-1~bddeb_all.deb
cloud-init clean --logs --reboot
# scp /var/log/cloud-init.log out
```
- [x] ubuntu/bionic [log](https://github.com/canonical/cloud-init/files/9522215/bionic.log)
- [x] ubuntu/focal [log](https://github.com/canonical/cloud-init/files/9522214/focal.log)
- [x] ubuntu/jammy [log](https://github.com/canonical/cloud-init/files/9522213/jammy.log)
- [x] upstream debian [log](https://github.com/canonical/cloud-init/files/9523336/debian.log)

<details>
<summary>The debian test passed, however.... 
</summary>

**tl;dr** debian packages apparmor profiles separate from individual packages (`apparmor-profiles`), and doesn't enable the dhclient apparmor profile by default on ec2.

Not only is the profile not installed by default, but it appears that I cannot even get the profile to load currently.
```
holmanb@ip-172-31-24-126:~$ sudo /usr/sbin/aa-enforce -d /usr/share/apparmor/extra-profiles/ dhclient

ERROR: Include file /usr/share/apparmor/extra-profiles/tunables/global not found
holmanb@ip-172-31-24-126:~$ dpkg -S /usr/share/apparmor/extra-profiles/sbin.dhclient
apparmor-profiles: /usr/share/apparmor/extra-profiles/sbin.dhclient
```
Since this error occurred, I couldn't actually load the profile to test it. So for all intents and purposes this test passed without apparmor, and didn't appear testable with apparmor.

Test Process:
-------------
Debian 11 Buster, the default image on EC2 is running a 20.x version of cloud-init. To ensure cloud-init dependencies were part of the image, I upgraded to bookworm (cloud-init 22.2) and did a test `cloud-init clean --logs --reboot` - there were some warnings, but nothing that looked like it would get in the way of the test.

Then I installed my test package, did a `cloud-init clean --reboot` and grabbed the logs.
<br>
</details>

Regarding Selinux:
------------------
Worth noting is that db86753f81af7382615 proposed by Redhat works around noexec issues by just calling the original binary. I'm hoping that this means that Selinux policies allow writing to `/run/dhclient.*` as well. Supporting this theory, I'm hopeful that [this might be the Selinux policy that enables that](https://github.com/SELinuxProject/refpolicy/blob/3a22e9279c1e80b58f35a0a77a8d6f7d164a0679/policy/modules/system/sysnetwork.fc#L83)?

### CODE REVIEW

File: tests/unittests/net/test_dhcp.py [hunk 1]
```diff
@@ -544,16 +511,15 @@ def test_dhcp_discovery_run_in_sandbox(self, m_subp, m_kill, m_getppid):
     @mock.patch("cloudinit.net.dhcp.util.get_proc_ppid")
     @mock.patch("cloudinit.net.dhcp.os.kill")
     @mock.patch("cloudinit.net.dhcp.subp.subp")
-    def test_dhcp_discovery_outside_sandbox(self, m_subp, m_kill, m_getppid):
-        """dhcp_discovery brings up the interface and runs dhclient.
-
-        It also returns the parsed dhcp.leases file generated in the sandbox.
-        """
+    @mock.patch("cloudinit.util.wait_for_files", return_value=False)
+    def test_dhcp_discovery_setup_and_run_dhclient(
```
Comment (TheRealFalcon) [inline tests/unittests/net/test_dhcp.py:515]: I think this test is now a duplicate of the test above it. If so, let's remove it.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Main change looks good, just a test issue.

**Comment by holmanb** [comment]
Hey @otubo I see you authored https://github.com/canonical/cloud-init/commit/db86753f81af73826158c9522f2521f210300e2b, which this PR affects. I'm pinging to see if you have any thoughts on the RHEL side.

This resolves what was an odd workaround with apparmor, which at some point caused issues on RHEL for noexec filesystems.

Considering that the fail-over introduced in db86753f81af73826158c9522f2521f210300e2b is to call the original binary, and in BZ [#1857309](https://bugzilla.redhat.com/show_bug.cgi?id=1857309) it was proposed to disable this feature in RHEL, as well as my selinux comments in the PR description, I _think_ we're safe to make this change.

Thoughts?

**Comment by holmanb** [comment]
@TheRealFalcon - I'm done testing this for now. I added notes and some logs for each of the tests I did. Only hiccup was with Debian (Debian doesn't appear to install this profile by default and my attempts to manually install and enable the profile failed).

**Comment by holmanb** [comment]
@rjschwei - Just an informational ping about this change.

I don't know whether apparmor `dhclient` profiles are enabled on cloud SLES/opensuse images. I wasn't planning on lifting pid/lease file definitions into distros, but it could be done if required by distro differences.

**Review by TheRealFalcon — APPROVED** [review]
LGTM. Let's give the people you pinged a few days to respond though.
